plugins/hls-semantic-tokens-plugin/src/Ide/Plugin/SemanticTokens/Query.hs

```diff
@@ -0,0 +1,135 @@
+                                                     mapMaybe)
+import qualified Data.Set                           as S
+import qualified Data.Set                           as Set
+import           Data.Text                          (Text)
+import           Development.IDE                    (HieKind,
+                                                     realSrcSpanToRange)
+import           Development.IDE.GHC.Compat
+import           Generics.SYB                       (mkQ)
+import           Ide.Plugin.SemanticTokens.Mappings
+import           Ide.Plugin.SemanticTokens.Types    (HsSemanticTokenType,
+                                                     NameSemanticMap)
+import           Language.LSP.Protocol.Types
+import           Prelude                            hiding (span)
+
+------------------------------------------------
+
+-- * extract names from RenamedSource
+
+------------------------------------------------
+
+nameToCollect :: Name -> NameSet
+nameToCollect = unitNameSet
+
+nameGetter :: RenamedSource -> NameSet
+nameGetter = everything unionNameSet (emptyNameSet `mkQ` nameToCollect)
+
+---------------------------------------------------------
+
+-- * extract semantic map from HieAst for local variables
+
+---------------------------------------------------------
+
+mkLocalNameSemanticFromAst :: HieKind a -> RefMap a -> NameSemanticMap
+mkLocalNameSemanticFromAst hieKind rm = mkNameEnv (mapMaybe (nameNameSemanticFromHie hieKind rm) ns)
+  where
+    ns = rights $ M.keys rm
+
+nameNameSemanticFromHie :: forall a. HieKind a -> RefMap a -> Name -> Maybe (Name, HsSemanticTokenType)
+nameNameSemanticFromHie hieKind rm ns = do
+  st <- -- traceShow ("to find Name", showName ns) $
+    nameSemanticFromRefMap rm ns
+  return
+    -- \$ traceShow (showName ns, st)
+    (ns, st)
+  where
+    nameSemanticFromRefMap :: RefMap a -> Name -> Maybe HsSemanticTokenType
+    nameSemanticFromRefMap rm' name' = do
+      spanInfos <- -- traceShow ("getting spans:", nameString) $
+        Map.lookup (Right name') rm'
+      let infos = S.unions $ map (identInfo . snd) spanInfos
+      let typeTokenType = fmap (typeSemantic hieKind) $ listToMaybe $ mapMaybe (identType . snd) spanInfos
+      let contextInfoTokenType = contextInfosMaybeTokenType infos
+      maximum <$> NE.nonEmpty (catMaybes [typeTokenType, contextInfoTokenType])
+
+    contextInfosMaybeTokenType :: Set.Set ContextInfo -> Maybe HsSemanticTokenType
+    contextInfosMaybeTokenType details = case NE.nonEmpty $ Set.toList details of
+      Just infos -> maximum $ NE.map infoTokenType infos
+      Nothing    -> Nothing
+
+-----------------------------------
+
+-- * extract location from HieAST a
+
+-----------------------------------
+
+hieAstSpanNames :: NameSet -> HieAST a -> [(Range, Name)]
+hieAstSpanNames nameSet ast =
+  if null (nodeChildren ast)
+    then getIds ast
+    else concatMap (hieAstSpanNames nameSet) (nodeChildren ast)
```

Comment:
Here's a place where it would be better to be using a map. We're doing a lot of list concatenation, and generally we're concatenating the same list prefix `n` times for a tree of depth `n`, so I think this is quadratic in the depth of the tree. Better to use a `Map` or `DList` or something with better concat performance.